### PR TITLE
Fix unable to sync json column with value is an array

### LINKED ISSUE #44459: Can't sync array columns
### Describe the bug

Unable to sync jsonb columns with value is an array

### To Reproduce

1. Use MB with appdb is PG
2. Create a postgres db with this table
```sql
CREATE TABLE my_table (
    id serial PRIMARY KEY,
    data jsonb
);

INSERT INTO my_table (data)
VALUES ('[1, 2, 3]'::jsonb);
```

3. Try to sync the created DB
4. Go to table metadata, you won't see this table has any fields

### Expected behavior

The table should be synced correctly with 2 columns: id, data

### Logs
```
2024-06-20 08:28:18,784 WARN sync.util :: Error checking if Fields ("data" "id" "data") need to be created or reactivated
clojure.lang.ExceptionInfo: ERROR: duplicate key value violates unique constraint "idx_uniq_field_table_id_parent_id_name_2col"
  Detail: Key (table_id, name)=(22, data) already exists. {:toucan2/context-trace [["execute SQL with class com.mchange.v2.c3p0.impl.NewProxyConnection" {:toucan2.jdbc.query/sql-args ["INSERT INTO \"metabase_field\" (\"description\", \"database_type\", \"semantic_type\", \"table_id\"
, \"coercion_strategy\", \"name\", \"updated_at\", \"effective_type\", \"nfc_path\", \"parent_id\", \"database_partitioned\", \"database_is_auto_increment\", \"json_unfolding\", \"position\", \"visibility_type\", \"display_name\", \"database_position\", \"database_required\", \"created_at\", \"base_type\") VALUES (NULL, ?, ?, ?, NULL, ?, NOW(), ?, NULL, NULL, NULL, FALSE, TRUE, ?, ?, ?, ?, FALSE, NOW(), ?), (NULL, ?, ?, ?, NULL, ?, NOW(), ?, NULL, NULL, NULL, TRUE, FALSE, ?, ?, ?, ?, FALSE, NOW(), ?), (NULL, ?, NULL, ?, NULL, ?, NOW(), ?, ?, NULL, NULL, FALSE, F
ALSE, ?, ?, ?, ?, FALSE, NOW(), ?)" "jsonb" "type/SerializedJSON" 22 "data" "type/JSON" 1 "details-only" "Data" 1 "type/JSON" "serial" "type/PK" 22 "id" "type/Integer" 0 "normal" "ID" 0 "type/Integer" "text" 22 "data" "type/Array" "[\"data\"]" 0 "normal" "Data" 0 "type/Array"]}] ["resolve connection" {:toucan2.connection/connectable org.postgresql.jdbc.PgConnection}] ["resolve connection" {:toucan2.connection/connectable nil}] {:toucan2.pipeline/rf #function[clojure.core/map/fn--5931/fn--5932]} ["with compiled query" {:toucan2.pipeline/compiled-query ["INSERT IN
TO \"metabase_field\" (\"description\", \"database_type\", \"semantic_type\", \"table_id\", \"coercion_strategy\", \"name\", \"updated_at\", \"effective_type\", \"nfc_path\", \"parent_id\", \"database_partitioned\", \"database_is_auto_increment\", \"json_unfolding\", \"position\", \"visibility_type\", \"display_name\", \"database_position\", \"database_required\", \"created_at\", \"base_type\") VALUES (NULL, ?, ?, ?, NULL, ?, NOW(), ?, NULL, NULL, NULL, FALSE, TRUE, ?, ?, ?, ?, FALSE, NOW(), ?), (NULL, ?, ?, ?, NULL, ?, NOW(), ?, NULL, NULL, NULL, TRUE, FALSE, ?
, ?, ?, ?, FALSE, NOW(), ?), (NULL, ?, NULL, ?, NULL, ?, NOW(), ?, ?, NULL, NULL, FALSE, FALSE, ?, ?, ?, ?, FALSE, NOW(), ?)" "jsonb" "type/SerializedJSON" 22 "data" "type/JSON" 1 "details-only" "Data" 1 "type/JSON" "serial" "type/PK" 22 "id" "type/Integer" 0 "normal" "ID" 0 "type/Integer" "text" 22 "data" "type/Array" "[\"data\"]" 0 "normal" "Data" 0 "type/Array"]}] ["with built query" {:toucan2.pipeline/built-query {:insert-into [:metabase_field], :values ((toucan2.instance/instance :model/Field {:description nil, :database_type "jsonb", :semantic_type "type/S
erializedJSON", :table_id 22, :coercion_strategy nil, :name "data", :updated_at [:metabase.util.honey-sql-2/typed :%now {:database-type "timestamptz"}], :effective_type "type/JSON", :nfc_path nil, :parent_id nil, :database_partitioned nil, :database_is_auto_increment false, :json_unfolding true, :position 1, :visibility_type "details-only", :display_name "Data", :database_position 1, :database_required false, :created_at [:metabase.util.honey-sql-2/typed :%now {:database-type "timestamptz"}], :base_type "type/JSON"}) (toucan2.instance/instance :model/Field {:des
cription nil, :database_type "serial", :semantic_type "type/PK", :table_id 22, :coercion_strategy nil, :name "id", :updated_at [:metabase.util.honey-sql-2/typed :%now {:database-type "timestamptz"}], :effective_type "type/Integer", :nfc_path nil, :parent_id nil, :database_partitioned nil, :database_is_auto_increment true, :json_unfolding false, :position 0, :visibility_type "normal", :display_name "ID", :database_position 0, :database_required false, :created_at [:metabase.util.honey-sql-2/typed :%now {:database-type "timestamptz"}], :base_type "type/Integer"})
(toucan2.instance/instance :model/Field {:description nil, :database_type "text", :semantic_type nil, :table_id 22, :coercion_strategy nil, :name "data", :updated_at [:metabase.util.honey-sql-2/typed :%now {:database-type "timestamptz"}], :effective_type "type/Array", :nfc_path "[\"data\"]", :parent_id nil, :database_partitioned nil, :database_is_auto_increment false, :json_unfolding false, :position 0, :visibility_type "normal", :display_name "Data", :database_position 0, :database_required false, :created_at [:metabase.util.honey-sql-2/typed :%now {:database-t
ype "timestamptz"}], :base_type "type/Array"}))}}] ["resolve connection" {:toucan2.connection/connectable metabase.db.connection.ApplicationDB}] ["resolve connection" {:toucan2.connection/connectable :default}] ["resolve connection" {:toucan2.connection/connectable nil}] ["with resolved query" {:toucan2.pipeline/resolved-query {}}] ["with parsed args" {:toucan2.pipeline/query-type :toucan.query-type/insert.pks, :toucan2.pipeline/parsed-args {:rows [{:description nil, :database_type "jsonb", :semantic_type :type/SerializedJSON, :table_id 22, :coercion_strategy ni
l, :name "data", :effective_type :type/JSON, :nfc_path nil, :parent_id nil, :database_partitioned nil, :database_is_auto_increment false, :json_unfolding true, :position 1, :visibility_type :details-only, :display_name "Data", :database_position 1, :database_required false, :base_type :type/JSON} {:description nil, :database_type "serial", :semantic_type :type/PK, :table_id 22, :coercion_strategy nil, :name "id", :effective_type :type/Integer, :nfc_path nil, :parent_id nil, :database_partitioned nil, :database_is_auto_increment true, :json_unfolding false, :posi
tion 0, :visibility_type :normal, :display_name "ID", :database_position 0, :database_required false, :base_type :type/Integer} {:description nil, :database_type "text", :semantic_type nil, :table_id 22, :coercion_strategy nil, :name "data", :effective_type :type/Array, :nfc_path [:data], :parent_id nil, :database_partitioned nil, :database_is_auto_increment false, :json_unfolding false, :position 0, :visibility_type :normal, :display_name "Data", :database_position 0, :database_required false, :base_type :type/Array}]}}] ["with model" {:toucan2.pipeline/model :
model/Field}] ["with unparsed args" {:toucan2.pipeline/query-type :toucan.query-type/insert.pks, :toucan2.pipeline/unparsed-args (:model/Field ({:description nil, :database_type "jsonb", :semantic_type :type/SerializedJSON, :table_id 22, :coercion_strategy nil, :name "data", :effective_type :type/JSON, :nfc_path nil, :parent_id nil, :database_partitioned nil, :database_is_auto_increment false, :json_unfolding true, :position 1, :visibility_type :details-only, :display_name "Data", :database_position 1, :database_required false, :base_type :type/JSON} {:descripti
on nil, :database_type "serial", :semantic_type :type/PK, :table_id 22, :coercion_strategy nil, :name "id", :effective_type :type/Integer, :nfc_path nil, :parent_id nil, :database_partitioned nil, :database_is_auto_increment true, :json_unfolding false, :position 0, :visibility_type
:normal, :display_name "ID", :database_position 0, :database_required false, :base_type :type/Integer} {:description nil, :database_type "text", :semantic_type nil, :table_id 22, :coercion_strategy nil, :name "data", :effective_type :type/Array, :nfc_path [:data], :parent_id nil, :da
tabase_partitioned nil, :database_is_auto_increment false, :json_unfolding false, :position 0, :visibility_type :normal, :display_name "Data", :database_position 0, :database_required false, :base_type :type/Array}))}]]}
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2725)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2412)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:502)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:419)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:194)
        at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:180)
        at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.execute(NewProxyPreparedStatement.java:67)
        at toucan2.jdbc.query$reduce_jdbc_query.invokeStatic(query.clj:40)
        at toucan2.jdbc.query$reduce_jdbc_query.invoke(query.clj:22)
        at toucan2.jdbc.pipeline$transduce_execute_with_connection_primary_method_java_sql_Connection_default_default.invokeStatic(pipeline.clj:19)
        at toucan2.jdbc.pipeline$transduce_execute_with_connection_primary_method_java_sql_Connection_default_default.invoke(pipeline.clj:9)
        at clojure.lang.AFn.applyToHelper(AFn.java:178)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:675)
        at clojure.core$partial$fn__5908.doInvoke(core.clj:2639)
        at clojure.lang.RestFn.applyTo(RestFn.java:146)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.invoke(RestFn.java:482)
        at toucan2.jdbc.pipeline$transduce_execute_with_connection_primary_method_java_sql_Connection_toucan_result_type_pks_default.invokeStatic(pipeline.clj:30)
        at toucan2.jdbc.pipeline$transduce_execute_with_connection_primary_method_java_sql_Connection_toucan_result_type_pks_default.invoke(pipeline.clj:22)
        at clojure.lang.AFn.applyToHelper(AFn.java:178)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:675)
        at clojure.core$partial$fn__5908.doInvoke(core.clj:2639)
        at clojure.lang.RestFn.applyTo(RestFn.java:146)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$apply.invoke(core.clj:662)
        at methodical.impl.combo.threaded$eval14915$fn__14916$fn__14917$fn__14924.invoke(threaded.clj:79)
        at methodical.impl.combo.threaded$reducer_fn$fn__14883$fn__14887.invoke(threaded.clj:23)
        at clojure.core.protocols$fn__8244.invokeStatic(protocols.clj:136)
        at clojure.core.protocols$fn__8244.invoke(protocols.clj:124)
        at clojure.core.protocols$fn__8204$G__8199__8213.invoke(protocols.clj:19)
        at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
        at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
        at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
        at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
        at clojure.core$reduce.invokeStatic(core.clj:6887)
        at clojure.core$reduce.invoke(core.clj:6869)
        at methodical.impl.combo.threaded$reducer_fn$fn__14883.invoke(threaded.clj:21)
        at clojure.core$comp$fn__5876.doInvoke(core.clj:2589)
        at clojure.lang.RestFn.applyTo(RestFn.java:146)
        at clojure.core$apply.invokeStatic(core.clj:675)
        at clojure.core$apply.doInvoke(core.clj:662)
        at clojure.lang.RestFn.invoke(RestFn.java:533)
        at methodical.impl.combo.threaded$combine_with_threader$fn__14893.doInvoke(threaded.clj:46)
        at clojure.lang.RestFn.applyTo(RestFn.java:151)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:675)
        at clojure.core$apply.doInvoke(core.clj:662)
        at clojure.lang.RestFn.invoke(RestFn.java:533)
        at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:65)
        at methodical.impl.standard$invoke_multifn.doInvoke(standard.clj:47)
        at clojure.lang.RestFn.invoke(RestFn.java:594)
        at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:199)
        at toucan2.pipeline$transduce_execute$with_connection_STAR___57873$with_transaction_STAR___57874.invoke(pipeline.clj:75)
        at toucan2.connection$bind_current_connectable_fn$fn__57458.invoke(connection.clj:104)
        at metabase.db.connection$do_with_transaction_primary_method_java_sql_Connection.invokeStatic(connection.clj:180)
        at metabase.db.connection$do_with_transaction_primary_method_java_sql_Connection.invoke(connection.clj:162)
        at clojure.lang.AFn.applyToHelper(AFn.java:165)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.invoke(RestFn.java:457)
        at clojure.core$partial$fn__5908.invoke(core.clj:2643)
        at clojure.lang.AFn.applyToHelper(AFn.java:160)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.invoke(RestFn.java:436)
        at methodical.impl.combo.threaded$eval14915$fn__14916$fn__14917$fn__14920.invoke(threaded.clj:71)
        at methodical.impl.combo.threaded$reducer_fn$fn__14883$fn__14887.invoke(threaded.clj:23)
        at clojure.lang.ArrayChunk.reduce(ArrayChunk.java:58)
        at clojure.core.protocols$fn__8244.invokeStatic(protocols.clj:136)
        at clojure.core.protocols$fn__8244.invoke(protocols.clj:124)
        at clojure.core.protocols$fn__8204$G__8199__8213.invoke(protocols.clj:19)
        at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
        at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
        at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
        at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
        at clojure.core$reduce.invokeStatic(core.clj:6887)
        at clojure.core$reduce.invoke(core.clj:6869)
        at methodical.impl.combo.threaded$reducer_fn$fn__14883.invoke(threaded.clj:21)
        at clojure.core$comp$fn__5876.invoke(core.clj:2588)
        at methodical.impl.combo.threaded$combine_with_threader$fn__14893.invoke(threaded.clj:44)
        at clojure.lang.AFn.applyToHelper(AFn.java:160)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.invoke(RestFn.java:436)
        at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invokeStatic(connection.clj:249)
        at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invoke(connection.clj:245)
        at clojure.lang.AFn.applyToHelper(AFn.java:165)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.invoke(RestFn.java:457)
```
### Information about your Metabase installation

```JSON
master
```

### Severity

P1

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by qnkhuat**
a json column with values is an array or an atom should not have nested fields

Closes #44459

### CODE REVIEW

File: src/metabase/driver/sql_jdbc/sync/describe_table.clj [hunk 1]
```diff
@@ -388,37 +388,45 @@
...
-            JsonToken/VALUE_STRING       (recur path key (assoc! res next-path (type-by-parsing-string (.getText p))))
-            JsonToken/FIELD_NAME         (recur path (.getText p) res)
-            JsonToken/START_OBJECT       (recur next-path key res)
-            JsonToken/END_OBJECT         (recur (cond-> path (seq path) pop) key res)
-            ;; We put top-level array row type semantics on JSON roadmap but skip for now
-            JsonToken/START_ARRAY        (do (.skipChildren p)
-                                             (recur path key (assoc! res next-path clojure.lang.PersistentVector)))
-            JsonToken/END_ARRAY          (recur path key res)))))))
+         ;; only parse if the value is an object, ignore arrays, strings, numbers, etc.
+         (and (zero? token-cnt)
```
Comment (qnkhuat) [inline src/metabase/driver/sql_jdbc/sync/describe_table.clj:398]: jackson doesn't have a peek method :(

File: src/metabase/driver/sql_jdbc/sync/describe_table.clj [hunk 2]
```diff
@@ -388,37 +388,45 @@
   Respects *nested-field-column-max-row-length*."
   [v path]
   (let [p (json-parser v)]
-    (loop [path (or path [])
-           key  nil
-           res  (transient {})]
-      (let [token     (.nextToken p)
-            next-path (cond-> path key (conj key))]
+    (loop [path      (or path [])
```
Comment (qnkhuat) [inline src/metabase/driver/sql_jdbc/sync/describe_table.clj:391]: The fix here is reverese https://github.com/metabase/metabase/pull/43812/files, then add a check to make sure we return nil if the value is not an object.

**Comment by qnkhuat** [comment]
[Thread](https://metaboat.slack.com/archives/C013VC33N95/p1718814222803569) for context

File: src/metabase/driver/sql_jdbc/sync/describe_table.clj [hunk 3]
```diff
@@ -388,37 +388,45 @@
...
+           keyy      nil
```
Comment (piranha) [inline src/metabase/driver/sql_jdbc/sync/describe_table.clj:392]: ehm, is this intentional?

Comment (qnkhuat) [reply src/metabase/driver/sql_jdbc/sync/describe_table.clj:392]: yea, I don't like the fact that it shadows function `key` :D.  we do this sometimes we keys like `type` and we use `ttype` instead.

maybe `kkey` is better?

File: src/metabase/driver/sql_jdbc/sync/describe_table.clj (hunk 1 above)
Comment (piranha) [reply src/metabase/driver/sql_jdbc/sync/describe_table.clj:398]: Why don't we check `(seq res)` here, seems like it should give the same result? Nothing collected and not a start of the object -> goodbye.

File: src/metabase/driver/sql_jdbc/sync/describe_table.clj (hunk 3 above)
Comment (piranha) [reply src/metabase/driver/sql_jdbc/sync/describe_table.clj:392]: ah! how do you like `field`? 😁

**Comment by qnkhuat** [comment]
need to backport to 49 too

**Review by sansolovyov — APPROVED** [review]
Except I'd love `and` to become `when` :)

**Comment by qnkhuat** [comment]
@metabase-bot backport release-x.49.x

**Pull request merged**